hax/hax/util.py

```diff
@@ -843,6 +853,135 @@ def get_node_encl_fid(self, node: str) -> Optional[Fid]:
+            LOG.debug('Setting enclosure=%s in KV with state=%s',
+                      encl_fid, value['state'])
+            self.kv.kv_put(encl['Key'], json.dumps(value))
+
+    @repeat_if_fails()
+    def set_ctrl_state(self, ctrl_fid: Fid, status: ServiceHealth) -> None:
+        # Example,
+        # {
+        #    "key": "m0conf/sites/0x5300000000000001:0x1/
+        #            racks/0x6100000000000001:0x2/encls/
+        #            0x6500000000000001:0x4/ctrls/
+        #            0x6300000000000001:0x5",
+        #    "value": "{\"state\": \"M0_NC_UNKNOWN\"}"
+        # }
+        ctrl_items = self.get_all_sites()
+        regex = re.compile(f'^m0conf\\/.*\\/ctrls/{ctrl_fid}$')
+        for ctrl in ctrl_items:
+            match_result = re.match(regex, ctrl['Key'])
+            if not match_result:
+                continue
+            value = json.loads(ctrl['Value'])
+            value['state'] = self.get_device_ha_state(status)
+            LOG.debug('Setting ctrl=%s in KV with state=%s',
+                      ctrl_fid, value['state'])
+            self.kv.kv_put(ctrl['Key'], json.dumps(value))
+
+    @repeat_if_fails()
+    def get_ctrl_state(self, obj_t: ObjT, fidk: int) -> int:
+        assert obj_t.value == ObjT.CONTROLLER.value
+        ctrl_fid = mk_fid(ObjT.CONTROLLER, fidk)
+        ctrl_items = self.get_all_sites()
+        regex = re.compile(f'^m0conf\\/.*\\/ctrls/{ctrl_fid}$')
+        for ctrl in ctrl_items:
+            match_result = re.match(regex, ctrl['Key'])
+            if not match_result:
+                continue
+            val = json.loads(ctrl['Value'])
+            state = val['state']
+            LOG.debug('Controller=%s state=%s', ctrl_fid, state)
+            if state in (m0HaObjState.M0_NC_ONLINE.name,
+                         m0HaObjState.M0_NC_TRANSIENT.name,
+                         m0HaObjState.M0_NC_FAILED.name):
+                return m0HaObjState.parse(state)
+        return m0HaObjState.M0_NC_ONLINE
+
+    @repeat_if_fails()
+    def get_encl_state(self, obj_t: ObjT, fidk: int) -> int:
+        assert obj_t.value == ObjT.ENCLOSURE.value
+        encl_fid = mk_fid(ObjT.ENCLOSURE, fidk)
+        encl_items = self.get_all_sites()
+        regex = re.compile(f'^m0conf\\/.*\\/encls/{encl_fid}$')
+
+        for encl in encl_items:
+            match_result = re.match(regex, encl['Key'])
+            if not match_result:
+                continue
+            val = json.loads(encl['Value'])
+            state = val['state']
+            LOG.debug('Enclosure=%s state=%s', encl_fid, state)
+            if state in (m0HaObjState.M0_NC_ONLINE.name,
+                         m0HaObjState.M0_NC_TRANSIENT.name,
+                         m0HaObjState.M0_NC_FAILED.name):
+                return m0HaObjState.parse(state)
+        return m0HaObjState.M0_NC_ONLINE
+
+    @repeat_if_fails()
+    def get_node_state(self, obj_t: ObjT, fidk: int) -> int:
+        assert obj_t.value == ObjT.NODE.value
```

Comment:
Hm, if we know for sure what `obj_t` SHOULD be, do we actually need to have it as a parameter?

Reply:
Indeed you are correct, but since get_sdev_state() uses obj_t and includes in the function param.. for consistency and function map `device_states_types{}` https://github.com/Seagate/cortx-hare/pull/1763/files#diff-1b0ad99a0bb8a23234ebf7e4be5891a9bdab039f1a1b64091a913e92dbf36387R536 .. added it as a param.

Thanks